Start capturing StackTrace-s in ErrorMonitor events.

xaionaro opened the pull request:
Solves: https://github.com/facebookincubator/go-belt/issues/11

# How it works

Now we have an interface "StackTrace" which is implemented by "PCs" and "Frames". And now default we capture PCs in an ErrorMonitor and use it as "StackTrace". We use "PCs" (and not "Frames") because it is the lightweight way.

xaionaro committed:
feat(errmon): Capture the StackTrace

How it works:

Now we have an interface "StackTrace" which is implemented by
"PCs" and "Frames". And now default we capture PCs in an ErrorMonitor
and use it as "StackTrace". We use "PCs" (and not "Frames") because
it is the lightweight way.

xaionaro committed:
bugfix(errmon/examples): Update Default func signature

xaionaro commented on tool/experimental/errmon/adapter/generic_error_monitor.go:
```diff
@@ -65,9 +65,13 @@ func (h *GenericErrorMonitor) newEvent(belt *belt.Belt, extraFields field.Abstra
 		ID: errmontypes.RandomEventID(),
 	}
 
+	// stack trace
+	ev.Exception.StackTrace = runtime.CallerStackTrace(h.CallerFrameFilter)
+
```
This line is the whole point of this PR.

alandiegosantos commented on pkg/runtime/stack_trace.go:
```diff
@@ -0,0 +1,165 @@
+	frameDepth := 1
+	for {
+		frame, ok := frames.Next()
+		result.WriteString(fmt.Sprintf("%d. %s:%d: %s\n", frameDepth, frame.File, frame.Line, frame.Function))
```
According to the iterator, it is possible that an empty value is returned by function, which might cause a panic in this line. Since the object is an interface, is it possible to add an extra check to evaluate if the object is not nil.

xaionaro replied:
Sorry, could you elaborate?

`frame` is not an interface, but a structure: [`runtime.Frame`](https://pkg.go.dev/runtime#Frame); while `frames` was checked against `nil` above and it is already not nil by this moment.

`frame` should not be empty according to [the example in the docs](https://pkg.go.dev/runtime#example-Frames).

u1f35c replied:
Should you not check `ok` before trying to access members of `frame`?

xaionaro replied:
No, this is bad naming. `ok` here means if `Next` should be called once more to get a new entry after that or this is the last entry.

xaionaro replied:
Updated the name, to make it more self-descriptive and less misleading. Now it is called `haveMore` instead of `ok`.

alandiegosantos commented on pkg/runtime/stack_trace.go:
```diff
@@ -0,0 +1,165 @@
+
+// Next implements FramesIterator.
+func (i *framesIterator) Next() (runtime.Frame, bool) {
+	frame := i.Frames[i.CurPosition]
```
if _Next()_ is called multiple times in sequence and the bool is not checked, this line can panic. In this case, I am not sure what is the desired 

Perhaps, something like this might solve this.
```suggestion
        if i.CurPosition >= len(i) {
            return <last_element|nil>, false
        }
	frame := i.Frames[i.CurPosition]
```

xaionaro replied:
It was intentional (due to performance reasons), but I took a look at [the standard `(*Frames).Next`](https://pkg.go.dev/runtime#Frames.Next) description, and you right:

> If it has already returned all call frames, Next returns a zero Frame.

Fixed. Thanks :)

xaionaro committed:
bug(runtime): Allow extra calls to *framesIterator.Next()

Adding a check if "CurPosition" is too high, to have the same
behavior as in standard (*runtime.Frames).Next():

> If it has already returned all call frames, Next returns a zero Frame.

xaionaro committed:
neat(runtime): rename 'ok' to 'haveMore'

'ok' usually implies if the main returned value could be used,
while here it actually means if there are more entries to be
fetched through more calls of Next().
